### PR TITLE
Monaco editor

### PR DESCRIPTION
**Pull request by MichaelPetrinolis**
Add monaco-editor (local and cdn) to resources, and use it as an editor for text and html fields. It depends on #8265 to load monaco loader and overcome the issue that only one anonymous define call per script file can be defined.

Related to #2345

### CODE REVIEW

**Comment by hishamco** [comment]
@agriffard finally seems someone take the initiative to introduce Monaco Editor

Thanks MichaelPetrinolis

**Comment by MichaelPetrinolis** [comment]
@hishamco thanx, is nice to have monaco in the CMS, especially in the resources, eg. we could use the diffEditor in the audit trail module to display differences and many other use cases. Don't know what to make a gif about, it is just another editor for TextField and HtmlField, with settings to define a json configuration for creating the editor.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Views/HtmlField-Monaco.Edit.cshtml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+@model OrchardCore.ContentFields.ViewModels.EditHtmlFieldViewModel
+
+@{
+    var settings = Model.PartFieldDefinition.GetSettings<HtmlFieldSettings>();
+    var monacoSettings = Model.PartFieldDefinition.GetSettings<TextFieldMonacoEditorSettings>();
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Views/HtmlField-Monaco.Edit.cshtml:5]:
```suggestion
    var monacoSettings = Model.PartFieldDefinition.GetSettings<HtmlFieldMonacoEditorSettings>();
```

**Comment by Skrypt** [comment]
Fun

image

**Comment by Skrypt** [comment]
image

Mobile seems to work but maybe needs some adjustments.

**Comment by hishamco** [comment]
@MichaelPetrinolis or @Skrypt this is really need a demo ;)

**Comment by hishamco** [comment]
@Skrypt did you try it in a light theme?

**Comment by Skrypt** [comment]
@hishamco Yeah, there are some small CSS glitches here and there but it mostly works.

image

**Comment by scleaver** [comment]
This is great!

**Comment by agriffard** [comment]
@MichaelPetrinolis Could you make a demo during next meeting?

**Comment by MichaelPetrinolis** [comment]
@agriffard I'll try to attend the meeting :)

**Comment by hishamco** [comment]
> @agriffard I'll try to attend the meeting :)

That's great!! or let someone else demo it in case you can't .. I can't wait for this one ;)

**Comment by hishamco** [comment]
@MichaelPetrinolis please fix the conflict

**Comment by MichaelPetrinolis** [comment]
@hishamco @agriffard I am afraid that due to personal reasons, I won't be able to attend. If someone else could demo, please do so. Sorry for the inconvenience.

**Comment by agriffard** [comment]
- Please add some documentation (at least in ContentFields): Config, settings, shortcuts, ...
- The Editor settings are displayed even if the Monaco editor mode is not selected.
- UI when you search:
image
- When I change the setttings `"language": "html"` to `"language": "javascript"` and I reedit, it is displaying `"language": "html"` again

**Comment by sebastienros** [comment]
Great demo about it today, thanks @MichaelPetrinolis

**Comment by agriffard** [comment]
Do you plan to also add it to HtmlBodyPart?

**Comment by MichaelPetrinolis** [comment]
@sebastienros sorry for not attending... FYI, I tested the json schema and works. Used a tool to extract Monaco configuration schema from source code and use it in the Monaco settings editor. The only issue I encountered was loading editor from CDN, CORS did not allow loading the above schema.
@agriffard I will fix the mentioned issues and add it to HtmlBodyPart. 

Thanx.

**Comment by agriffard** [comment]
Please solve the conflicts.

**Comment by agriffard** [comment]
Build error:
`Error: D:\a\OrchardCore\OrchardCore\test\OrchardCore.Tests\ResourceManagement\ResourceManagerTests.cs(334,21): error CS0111: Type 'ResourceManagerTests' already defines a member called 'RequireCircularDependenciesShouldThrowException' with the same parameter types [D:\a\OrchardCore\OrchardCore\test\OrchardCore.Tests\OrchardCore.Tests.csproj]`

**Comment by MichaelPetrinolis** [comment]
I added typescript-json-schema as a devDependency and use it to create the IStandaloneEditorConstructionOptions.json schema
```
.\node_modules\.bin\typescript-json-schema 'node_modules/monaco-editor/monaco.d.ts' 'monaco.editor.IStandaloneEditorConstructionOptions' -o wwwroot/Scripts/monaco/IStandaloneEditorConstructionOptions.json
```
this is how it looks when editing settings
image

I also fixed the css issue (there is still an animation with border-color when clicking to expand the replace section)

File: src/OrchardCore.Modules/OrchardCore.Resources/ResourceManifest.cs [hunk 2]
```diff
@@ -392,6 +402,35 @@ public void BuildManifests(IResourceManifestBuilder builder)
...
+                .SetDependencies("monaco-loader")
+                .SetVersion(monacoEditorVersion);
+
+            //manifest
+            //    .DefineScript("monaco-editor-nls")
+            //    .SetUrl("~/OrchardCore.Resources/Scripts/monaco/min/vs/editor/editor.main.nls.js", "~/OrchardCore.Resources/Scripts/monaco/dev/vs/editor/editor.main.nls.js")
+            //    .SetDependencies("monaco")
+            //    .SetVersion(monacoEditorVersion);
+
+            //manifest
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Resources/ResourceManifest.cs:429]: Leftover?

File: src/OrchardCore.Modules/OrchardCore.Resources/OrchardCore.Resources.csproj [hunk 3]
```diff
@@ -10,10 +10,32 @@
...
+
   <ItemGroup>
     <FrameworkReference Include="Microsoft.AspNetCore.App" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Include="wwwroot\Scripts\monaco\cdn.js">
+      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
+    </None>
+    <None Include="wwwroot\Scripts\monaco\dev.js">
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Resources/OrchardCore.Resources.csproj:28]: This confuses me? Do you need to do this?

**Review by deanmarcussen — COMMENTED** [review]
Got a merge conflict to resolve still.

And a couple of questions, but think now we've got the resource position sorted out, then this is pretty much there.

**Comment by MichaelPetrinolis** [comment]
@deanmarcussen thanx for checking, fixed merge issues and did some minor changes. VS2019 modifies .csproj when you copy/move items through the solution explorer, that's why the items in the project file :)

There is also an issue when using CDN, it does not load the IStandaloneEditorConstructionOptions json schema for intelisense

**Comment by deanmarcussen** [comment]
ok great @MichaelPetrinolis 

re this `IStandaloneEditorConstructionOptions` issue, I'm not sure what it does, or why there is a problem with it?

Maybe some options to get this pr moving forward and merged
- Remove the feature that I don't know what is from this pr, to be implemented later
- Fix it?
- If it's crucial, and isn't going to work form a cdn, do we just remove the cdn urls, and always run locally (not a killer, it is for the admin after all, rather than the front end)

is it critical, i.e. can we drop it, and figure out how to implement later, or do we have to have it?

**Comment by agriffard** [comment]
As dean said, tell us if you can finish the PR without the `IStandaloneEditorConstructionOptions` if it is a blocker so it can be merged.

Please fix the conflict.

If you add some docs, you can make a link to this demo: https://www.youtube.com/watch?v=lnjdRildsL8

**Comment by MichaelPetrinolis** [comment]
@agriffard sure, i will proceed as soon as i get some time. The IStandaloneEditorConstructionOptions is a json schema that is used for providing intellisense when you edit the Monaco initialization options in Admin. The issue is when Monaco is served by CDN, it cannot load the above json schema due to CORS permissions. So intellisense does not work when using CDN resources.

**Comment by agriffard** [comment]
Please fix the conflicts (ex: Resources moved to ResourceManagementOptionsConfiguration).

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/HtmlFieldMonacoEditorSettingsDriver.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                var settings = partFieldDefinition.GetSettings<HtmlFieldMonacoEditorSettings>();
+                if (string.IsNullOrWhiteSpace(settings.Options))
+                {
+                    settings.Options = JsonConvert.SerializeObject(new { automaticLayout = true, language = "html" }, Formatting.Indented);
+                }
+                model.Options = settings.Options;
+            })
+            .Location("Editor");
+        }
+
```
Comment (agriffard) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/HtmlFieldMonacoEditorSettingsDriver.cs:37]: Remove extra line.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/TextFieldMonacoEditorSettingsDriver.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                var settings = partFieldDefinition.GetSettings<TextFieldMonacoEditorSettings>();
+                if (string.IsNullOrWhiteSpace(settings.Options))
+                {
+                    settings.Options = JsonConvert.SerializeObject(new { automaticLayout = true, language = "html" }, Formatting.Indented);
+                }
+                model.Options = settings.Options;
+            })
+            .Location("Editor");
+        }
+
```
Comment (agriffard) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/TextFieldMonacoEditorSettingsDriver.cs:36]: Remove extra line.

**Comment by agriffard** [comment]
Can you please add a new eol for each file that lacks of it?

File: src/OrchardCore.Modules/OrchardCore.Resources/package.json [hunk 6]
```diff
@@ -12,11 +12,15 @@
     "jquery-resizable-dom": "0.35.0",
     "jquery.easing": "1.4.1",
     "js-cookie": "2.2.1",
+    "monaco-editor": "^0.22.3",
```
Comment (agriffard) [inline src/OrchardCore.Modules/OrchardCore.Resources/package.json:15]: Please update to https://www.npmjs.com/package/monaco-editor/v/0.23.0 and version in docs/resources/libraries/Readme.md

**Comment by agriffard** [comment]
The shortcode modal does not seem to have a way to appear in the Part (no button as in other editors).

If you manage to make it work, please add it for the field editor too.

/cc @deanmarcussen

**Comment by agriffard** [comment]
Thank you but, please, choose another shortcut than `Ctrl+c` so we can still copy some text.

**Comment by MichaelPetrinolis** [comment]
@agriffard haha, how did this happen! What about alt + c? Any suggestions are welcome

**Comment by agriffard** [comment]
> @agriffard haha, how did this happen! What about alt + c? Any suggestions are welcome

`Alt+c` already opens a 'Match case' dialog.

**Comment by agriffard** [comment]
Can you please merge dev to fix the conflicts?

**Review by agriffard — APPROVED** [review]
LGTM

**Review by deanmarcussen — APPROVED** [review]
Huge pr. Thanks @MichaelPetrinolis

**Comment by jptissot** [comment]
Thanks for this ! Kudos. 

I have a question: Do we need all these languages ?

**Comment by sebastienros** [comment]
We should only have the minified versions of the files. I don't think it makes sense to redistribute all sources. We can depend on npm but only copy the min files.

**Comment by Skrypt** [comment]
I was trying to find Liquid in this list of languages but it's not supported it requires an extension in vs code so it's the same for the editor here.

**Pull request merged**